polygeist/lib/Dialect/Polygeist/Transforms/ArgumentPromotion.cpp

```diff
@@ -0,0 +1,385 @@
+
+  funcOp.setFunctionType(newFuncType);
+  funcOp.setAllArgAttrs(newArgAttrs);
+
+  LLVM_DEBUG(llvm::dbgs() << "New Callee:\n" << funcOp << "\n\n");
+}
+
+void Candidate::replaceArgumentWith(unsigned pos, Region &callableRgn,
+                                    const SmallVector<Value> &newArgs,
+                                    SmallVector<Attribute> &newArgAttrs) const {
+  assert(!newArgs.empty() && "Expecting a non-empty vector");
+
+  Value origArg = callableRgn.getArgument(pos);
+
+  for (unsigned offset = 0; offset < newArgs.size(); ++offset) {
+    callableRgn.insertArgument(pos + offset, newArgs[offset].getType(),
+                               callableRgn.getLoc());
+    // TODO: add attribute NoAlias on the new argument.
+    newArgAttrs.push_back(Attribute());
+  }
+
+  // Replace uses of the original argument with the new arguments injected.
+  replaceUsesOfArgument(origArg, pos, callableRgn);
+}
+
+void Candidate::replaceUsesOfArgument(Value origArg, unsigned pos,
+                                      Region &callableRgn) const {
+  for (OpOperand &use : origArg.getUses()) {
+    assert(isa<polygeist::SubIndexOp>(use.getOwner()) &&
+           "Expecting a subindex operation");
+
+    // The use must be a polygeist::SubIndexOp operation with a constant
+    // index, remove it and rewrite its users.
+    auto subIndexOp = cast<polygeist::SubIndexOp>(use.getOwner());
+    TypedValue<IndexType> index = subIndexOp.getIndex();
+    auto indexOp = index.getDefiningOp<arith::ConstantIndexOp>();
+    assert(indexOp && "Must have a constant index");
+
+    int64_t offset = indexOp.value();
+    BlockArgument newArg = callableRgn.getArgument(pos + offset);
+    subIndexOp.replaceAllUsesWith(newArg);
+    subIndexOp.erase();
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// ArgumentPromotionPass
+//===----------------------------------------------------------------------===//
+
+void ArgumentPromotionPass::runOnOperation() {
+  SmallVector<Candidate> candidates;
+  collectCandidates(candidates);
+
+  for (Candidate &cand : candidates)
+    cand.transform();
+}
+
+void ArgumentPromotionPass::collectCandidates(
+    SmallVectorImpl<Candidate> &candidateCalls) {
+  ModuleOp module = getOperation();
+  SymbolTableCollection symTable;
+  SymbolUserMap userMap(symTable, module);
+
+  // Collect candidate call operations in GPU kernels.
+  module->walk([&](gpu::GPUFuncOp gpuFuncOp) {
+    LLVM_DEBUG({
+      llvm::dbgs()
+          << "Analyzing: "
+          << cast<SymbolOpInterface>(gpuFuncOp.getOperation()).getNameAttr()
+          << "\n";
+    });
+
+    gpuFuncOp->walk([&](CallOpInterface callOp) {
+      // We are only interested in non-recursive tail calls.
+      if (!callOp->getBlock()->hasNoSuccessors() ||
+          Candidate::isRecursiveCall(callOp))
+        return WalkResult::advance();
+
+      // The first operand of the call must not be used by any other
```

Comment:
What's the problem when the first operand is used by other operation?

Reply:
I was trying to restrict the number of cases this transformation kicks in.

Reply:
ok removed the restriction